### PR TITLE
feat(unstable-pagination): add refactored experimental pagination component

### LINKED ISSUE #4365: Page selector is not suitable for tables with many pages
## Title line template: [Title]: Brief description
The page selector on https://pages.github.ibm.com/security/carbon-addons-security/branch/v2/?path=/story/components-datatablepagination--pagination-data-table component is not suitable for tables with many pages

## What package(s) are you using?

- [ ] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

The page selector on https://pages.github.ibm.com/security/carbon-addons-security/branch/v2/?path=/story/components-datatablepagination--pagination-data-table component that we are using in our app is unsuitable when there are many pages in the table - would it be possible to use a different component?

**Expected behavior -**

User should be able to enter a number to select the page via, for example, a TextInput or a NumberInput component

**Actual behavior -**

The control is a Select inside the Pagination component, and when there are many pages the it is displayed outside the confines of the browser page, and user has to scroll up and down to find the required page and then select it. This is not an acceptable User Experience.

## Steps to reproduce the issue

1. Create around 650 entries in the data table pagination table
2. Select 10 items per page
3. Click on the page selector and observe it's behaviour when trying to select a page, say page 60

**Affected browsers**
All browsers, but worse on Chrome (as shown in screenshot)

image

### LINKED ISSUE #5129: Pagination "totalItems" bigger then 2000000 is blocking whole application
## What package(s) are you using?

- [x] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

> Describe in detail the issue you're having.

In my application, I'm using carbon react component "pagination".  In the pagination component I'm using "totalItems" property I'm  combination with itemRangeText function. When number of totalitems is very big in my situation close to 2 000 000 whole application is blocked.

> Is this issue related to a specific component?

"Pagination" Component

> What did you expect to happen? What happened instead? What would you like to
> see changed?

> What browser are you working in?
Firefox and Chrome

> What version of the Carbon Design System are you using?

"carbon-components": "^10.7.0",
"carbon-components-react": "^7.7.0",

> What offering/product do you work on? Any pressing ship or release dates we
> should be aware of?

## Steps to reproduce the issue

It is easy to reproduce even on official Stroybook

1. Open link with Pagination component in Storybook
http://react.carbondesignsystem.com/?path=/story/pagination--pagination
2. Click on "Knobs" tab
3. In the input field "Total number of items (totalitems)" put number 2000 000.
4. leave the input field and everything is blocked.
.

> Please create a reduced test case in CodeSandbox
>
> - Style and vanilla JS:
>   https://codesandbox.io/s/github/carbon-design-system/carbon/tree/master/packages/components/examples/codesandbox
> - React:
>   https://codesandbox.io/s/github/carbon-design-system/carbon/tree/master/packages/react/examples/codesandbox

## Workaround

It is possible to use property "pageInputDisabled" in combination with function "itemText" and just manually put total items.

### PR DESCRIPTION
**Pull request by jendowns**
Closes #5129 
Closes #4365 

This PR has the `Unstable_Pagination` component, a refactored experimental version of `Pagination` component that is written with a "building blocks" approach, allowing users to add either an inner `PageSelector` (traditional select) ~or `PageInput` (a number input),~ or leave out those `children` entirely. **The component also includes icon-only buttons with tooltips.**

This component is implemented in our downstream library `@carbon/ibm-security`, and I copied almost everything from there (changing prefixes and so forth, of course).

### Changelog

**New**

- add `Unstable_Pagination` component with demo and styles
- add `PageSelector`
- the **pagiantion buttons now have tooltips!!** 🎉
- ~add `PageInput`~ Removed in this PR, to wait for design exploration in #5546

### TODO: Tests

I couldn't copy over our tests because we use react testing library for our version of this component -- its' super useful because we have the user event package as well to check tab order etc.

Are you all open to that, or would you rather I rewrite tests in your current structure/style?

### Testing / Reviewing

Check out the storybook demo & documentation.

### CODE REVIEW

**Comment by jendowns** [comment]
@designertyler This PR is a followup to our conversation in #4365 --

I have an example with a number input added here: https://deploy-preview-5485--carbon-components-react.netlify.com/?path=/story/unstable-pagination--with-a-page-input
How should it look though?
And where should the error state appear? (right now I have hidden the "invalid"/"error" text for the input, and all you see if a red outline if you type in letters)

**Comment by joshblack** [comment]
@jendowns for the use-case of tables with a large number of pages, what use-case or flow are we considering where they would like to select a specific page in a list of 200,000 pages? 🤔

**Review by joshblack — COMMENTED** [review]
Left a question above, curious to hear what you think 👀

**Comment by jendowns** [comment]
@joshblack Here are the use cases I'm seeing...

- Super large number of pages (10k, 200k, etc) -- don't use `PageSelector`. Maybe use `PageInput` *if you want* and if you are getting data from backend with each page change (othewrise that's... a lot of data lol). 
- 100-ish pages -- could use `PageSelector` and pass in `size={10}` to limit options list height (since `PageSelector` has a spread operator). Or could use `PageInput`.

I think the key in both cases is to pick the tool that works best, whether it's a select, a number input, or nothing at all. 👍

**Comment by jendowns** [comment]
OH @joshblack I just realized I kinda missed the point of your question here. 😅 

Yeah I don't see why there's a need to jump pages at all, honestly. It seems like you would have to know EXACTLY where you want to go?

But these are just options 🤷‍♀ As said in my previous comment, it's a set of tools that someone can use depending on their situation. If they have 200k pages they should seriously consider NO `children` for this pagination component and instead provide ways for people to filter/search.

**Comment by joshblack** [comment]
@jendowns yeah, I think that's 100% along the lines of what I was thinking for this too. It can seem like at that number of pages this feature isn't really meant for your use-case/users.

**Comment by jendowns** [comment]
@joshblack 200k is defintiely an extreme example. What #4365 highlights an issue we have in product *right now* where there's a reasonable number of pages (80-something, I think) and the page selector is super long in Chrome.

Try it out here in Chrome: https://react.carbondesignsystem.com/?path=/story/pagination--pagination
Set the `totalItems` to like 500 or something.

**Comment by designertyler** [comment]
Using the number input in place means we have two sets of buttons with the same function right next to each other.

Screen Shot 2020-03-02 at 10 07 25 AM

There's no existing example of an input in a toolbar like this so we'd have to explore ways to do the error message or limit the ability for the user to cause an error. I'm also not sure if we want to use ui-03 for the field or if we'd want to use something to the mobile number input that uses field-01.

I'll get with the team to see how we want to establish these types of entry fields.

**Comment by jendowns** [comment]
Thanks @designertyler!

If you prefer, I can remove the `PageInput` subcomponent (i.e., the inner number input) from this PR for now, while you explore the button & error state questions? (And then when you are finished with the design explorations, I can propose a followup PR, separate from this one.)

That way, I can still get a review on the rest of the code with the hopes of providing good alternatives / paths for people to resolve issues like #5129 & #4365? (That is, with this PR minus the number input, they'd be able to just leave out any type of page selector to avoid causing their browser to crash. OR pass a size attribute to the page selector to limit the size of the options list, if they don't have thousands of pages.)

CC @joshblack @tw15egan

**Comment by designertyler** [comment]
@jendowns 

Breaking out the page input makes sense. I created a design issue to track requirements and explorations for a pagination number input here https://github.com/carbon-design-system/carbon/issues/5546

**Comment by jendowns** [comment]
Thank you @designertyler! 🎉 

@joshblack @tw15egan I have removed the `PageInput` subcomponent completely from this PR.

So now the PR just adds the experimental `Pagination` refactor + a `PageSelector` subcomponent.

To reiterate what was said above, this refactor is different from the current `Pagination` component because of its flexibility. You can leave out the `PageSelector` completely to deal with super huge data sets (see: #4365) Or, since `PageSelector` is its own thing that takes its own props, you can pass in a `size` attribute to reduce the size of the `<select />` options list if that makes sense for you (which would solve #5129 ).

And because of the flexibility, once the design exploration is finished in #5546 we can come back and add a design-approved number input that could be used instead of the `PageSelector`.

This refactor also uses the icon-only `Button` so it has tooltips. 👍

**Comment by tw15egan** [comment]
Okay been trying to follow along with this PR, thank you for taking the time to work on this @jendowns! I like the idea that `PageSelector` is broken out into its own component for more flexibility. 

~When you say `size` prop, I don't see any in `PageSelector`. Do you mean they can modify the `pageSize` prop on the `Unstable_Pagination` to reduce the number of items shown on the `PageSelector` itself?~

Actually, reading further into one of your comments, I see that you're talking about the native HTML `size` property on `select` elements (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select)

It seems like OUR prop `size` is probably overriding the native functionality of this since we're passing in a string, and it is expecting a number. Seems like we've run into another issue with this 

Screen Shot 2020-03-09 at 2 09 01 PM

Did not realize there would be a collision there when we added the `size` prop, so we've definitely got to change that. Another item to add to the large v11 make prop names consistent epic.

However, it seems like `size` can only be used if the `multiple` attribute is also set? And when it is the styles seem to break...

Definitely seems like a number input would be the best solution for this, so it will be great to see this once `PageSelector` can be used as a select OR a number input. 


I also noticed one tiny visual bug, buttons seem to jump a bit when focused.

pagination

**Comment by jendowns** [comment]
Thanks @tw15egan I will fix that button issue 👀 

With regards to the `size` attribute for `select` HTML elements, that is unfortunate that it is now being overridden by a prop. (But I'm glad you all are going to change the colliding prop name in the next major release?). While `size` is often used with `multiple`, I've never had issues using it for non-multi `select`s? 🤷‍♀ And while styles break a little bit, in my closed PR https://github.com/carbon-design-system/carbon/pull/4378 -- it was just a couple style updates required to get it looking "decent". I'm not suggesting it as a standard way, since that PR #4378 wasn't favored anyway, BUT just wanted to present it as one of many options that this PR could open up.

(EDIT: also please let me know if you have any more questions about the origins of this PR, or its history. I've been trying to tag previous related PRs and issues 😅 There's a LOT lol)

**Comment by tw15egan** [comment]
Ah okay! I may have been adding it wrong and didn't change any styles so that's probably why I was seeing issues 👍

**Comment by jendowns** [comment]
Hey @tw15egan sorry for the delay! I have fixed the jumpy focus border/outline that you saw when you reviewed. Could you take another look when you get a chance? Thank you! 🎉

**Comment by tw15egan** [comment]
@jendowns just another small style issue I noticed. Seems like the spacing between the chevrons are different. Also, the right select does not seem to have any hover styles.

pagination-2


Other than that, everything looks great!

**Comment by jendowns** [comment]
@joshblack @tw15egan I updated the hover styles for the "page selector" select element.

I'm not sure about the point about arrow spacing in the select inputs, however. They both have the same svg, the same dimensions, and the same padding:
Screen Shot 2020-05-12 at 1 28 57 PM
Screen Shot 2020-05-12 at 1 29 16 PM

**Comment by jendowns** [comment]
Hey @joshblack @tw15egan @designertyler did you all need anything from me with this PR?

**Review by designertyler — APPROVED** [review]
Visually it looks good. I think the chevron spacing was a result of there being 1 vs 2 digit numbers next to the chevron.

File: packages/react/src/components/Pagination/Unstable_Pagination/Pagination.js [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+import classnames from 'classnames';
+import { settings } from 'carbon-components';
+import { CaretRight24, CaretLeft24 } from '@carbon/icons-react';
+import Button from '../../Button';
+import Select from '../../Select';
+import SelectItem from '../../SelectItem';
+
+const { prefix } = settings;
+
+function Unstable_Pagination({
```
Comment (joshblack) [inline packages/react/src/components/Pagination/Unstable_Pagination/Pagination.js:19]:
```suggestion
function unstable_Pagination({
```

Will try to make sure this is in our style guide soon! Totally get if the `U` was meant for usage in React components, I think we would just recommend renaming the import if that's the case:

```js
import { unstable_Pagination as Pagination } from '...';
```

**Review by joshblack — APPROVED** [review]
Definitely down to bring this in under the unstable namespace 👀 So sorry for the delays @jendowns, thanks so much for being persistent here 🙏

File: packages/react/src/components/Pagination/Unstable_Pagination/Pagination.js (hunk 1 above)
Comment (jendowns) [reply packages/react/src/components/Pagination/Unstable_Pagination/Pagination.js:19]: @joshblack I can update it to `unstable_` in all locations (including file names) if that's what you prefer?

Comment (joshblack) [reply packages/react/src/components/Pagination/Unstable_Pagination/Pagination.js:19]: Yes please! The only reason we do this convention is to mirror FB's `unstable_` convention, e.g. `unstable_SuspenseList`. I'll make sure to add a reference to this in our style guide as well, sorry about the mixup

Comment (jendowns) [reply packages/react/src/components/Pagination/Unstable_Pagination/Pagination.js:19]: Yeah no worries @joshblack I probably misremembered the capitalization when I wrote this

One thing I just noticed though, as I started to change to `unstable_` -- the linter is not pleased with me & flags hooks usage as a violation of `eslint(react-hooks/rules-of-hooks)`:

Screen Shot 2020-06-01 at 10 13 35 AM

> React Hook "useState" is called in function "unstable_Pagination" that is neither a React function component nor a custom React Hook function. eslint(react-hooks/rules-of-hooks)

How do you want to proceed here?

**Review by tw15egan — APPROVED** [review]
Changes look good to me! Thanks for taking the time to add this in @jendowns

**Pull request merged**